chainntnfs/btcdnotify/btcd.go

```diff
@@ -290,80 +276,68 @@ out:
-					continue
+				// Lookup whether the transaction is already included in the
+				// active chain.
+				txConf, err := b.historicalConfDetails(msg.TxID)
+				if err != nil {
+					chainntnfs.Log.Error(err)
 				}
-
-				txid := *msg.txid
-				b.confNotifications[txid] = append(b.confNotifications[txid], msg)
+				b.txConfNotifier.Register(&msg.ConfNtfn, txConf)
 			case *blockEpochRegistration:
 				chainntnfs.Log.Infof("New block epoch subscription")
 				b.blockEpochClients[msg.epochID] = msg
 			}
 
-		case staleBlockHash := <-b.disconnectedBlockHashes:
-			// TODO(roasbeef): re-orgs
-			//  * second channel to notify of confirmation decrementing
-			//    re-org?
-			//  * notify of negative confirmations
-			chainntnfs.Log.Warnf("Block disconnected from main "+
-				"chain: %v", staleBlockHash)
-
-		case <-b.chainUpdateSignal:
-			// A new update is available, so pop the new chain
-			// update from the front of the update queue.
-			b.chainUpdateMtx.Lock()
-			update := b.chainUpdates[0]
-			b.chainUpdates[0] = nil // Set to nil to prevent GC leak.
-			b.chainUpdates = b.chainUpdates[1:]
-			b.chainUpdateMtx.Unlock()
-
-			currentHeight = update.blockHeight
-
-			newBlock, err := b.chainConn.GetBlock(update.blockHash)
-			if err != nil {
-				chainntnfs.Log.Errorf("Unable to get block: %v", err)
-				continue
-			}
+		case item := <-b.chainUpdates.ChanOut():
+			update := item.(*chainUpdate)
+			if update.connect {
+				if update.blockHeight != currentHeight+1 {
+					chainntnfs.Log.Warnf("Received blocks out of order: "+
+						"current height=%d, new height=%d",
+						currentHeight, update.blockHeight)
+				}
 
-			chainntnfs.Log.Infof("New block: height=%v, sha=%v",
-				update.blockHeight, update.blockHash)
-
-			b.notifyBlockEpochs(update.blockHeight,
-				update.blockHash)
-
-			newHeight := update.blockHeight
-			for i, tx := range newBlock.Transactions {
-				// Check if the inclusion of this transaction
-				// within a block by itself triggers a block
-				// confirmation threshold, if so send a
-				// notification. Otherwise, place the
-				// notification on a heap to be triggered in
-				// the future once additional confirmations are
-				// attained.
-				txSha := tx.TxHash()
-				b.checkConfirmationTrigger(&txSha, update, i)
-			}
+				currentHeight = update.blockHeight
+
+				rawBlock, err := b.chainConn.GetBlock(update.blockHash)
+				if err != nil {
+					chainntnfs.Log.Errorf("Unable to get block: %v", err)
+					continue
+				}
+
+				chainntnfs.Log.Infof("New block: height=%v, sha=%v",
+					update.blockHeight, update.blockHash)
+
+				b.notifyBlockEpochs(update.blockHeight, update.blockHash)
 
-			// A new block has been connected to the main
-			// chain. Send out any N confirmation notifications
-			// which may have been triggered by this new block.
-			b.notifyConfs(newHeight)
+				txns := btcutil.NewBlock(rawBlock).Transactions()
+				err = b.txConfNotifier.ConnectTip(update.blockHash,
```

Comment:
Admittedly I'm not totally familiar with how btcd callbacks fire, but if a reorg happens, I imagine that you'd probably get callbacks for new blocks and disconnected blocks around the same time. If the new block callback is processed first, won't this cause the block to get lost as the txConfNotifier doesn't know that a reorg has occurred, so will drop the update?

Reply:
`btcd` fires `disconnectblock` notifications in reverse order, and then fires `connectblock` for each new connected block in order. The same goes for `neutrino`, which intentionally replicates the behavior.